plugin/core/views.py

```diff
@@ -172,3 +175,42 @@ def text_document_range_formatting(view: sublime.View, region: sublime.Region) -
         "options": formatting_options(view.settings()),
         "range": region_to_range(view, region).to_lsp()
     })
+
+
+def minihtml(content: Optional[Union[str, dict]], view: sublime.View) -> str:
+    """ Content can be a string or MarkupContent """
+    if isinstance(content, str):
+        return text2html(content)
+    elif isinstance(content, dict):
+        value = content.get("value") or ""
+        kind = content.get("kind")
+        if kind == "markdown":
+            return mdpopups.md2html(view, value)
+        else:
+            # must be plaintext
+            return text2html(value)
+    else:
+        return ''
+
+
+def text2html(content: str) -> str:
+    content = html.escape(content).replace('\n', '<br>')
```

Comment:
I did some experimentation. To get whitespace indentation all we have to do is this:

```
content = html.escape(content).replace('\n', '<br>').replace(' ', '&nbsp;')
```
minihtml seems to respect multiple `&nbsp;` escapes after each other, so this works.
Please add some tests for `text2html` in test_views.py because it has subtle behavior as we've learned.

Reply:
This plaintext string:

```
first line\nsecond line\n    third line indented\n        more indent\nback on the first column again
```
is rendered as
![Screenshot from 2020-05-09 12-05-05](https://user-images.githubusercontent.com/2431823/81470858-6b32f600-91ed-11ea-8a0a-d2216c795d34.png)

Reply:
I forgot about tabs. We can do this for tabs:

```
html.escape(content).replace('\n', '<br>').replace(' ', '&nbsp;').replace('\t', '&nbsp;&nbsp;&nbsp;&nbsp;')
```

Reply:
I am now wondering if we should preserve space for plaintext content. 
Given we have content that is like this:
```
Specifies the width of the content area, padding area or border area (depending on 'box-sizing') of certain boxes.

Syntax: <viewport-length>{1,2}

MDN Reference: https://developer.mozilla.org/docs/Web/CSS/width
```

If we use `.replace(' ', '&nbsp;').replace('\t', '&nbsp;&nbsp;&nbsp;&nbsp;')`) we will see exactly the same formatting as in the code block above:

![aa](https://user-images.githubusercontent.com/22029477/81507406-81c87280-92fd-11ea-8482-ad89348fe79e.png)
But a horizontal scrollbar is shown.

Without it we will see:
![ab](https://user-images.githubusercontent.com/22029477/81507413-8db43480-92fd-11ea-85bc-d0db0f8a7b87.png)
The text is wrapped, but a vertical scrollbar is shown.

I would rather see a vertical scrollbar than a horizontal one.
What do you guys think?

Reply:
I don't like horizontal scrolling either.
So what was the problem with using triple backticks? Is it that it changes font look?

Reply:
Since we are desperate and coming up with crazy workarounds, maybe we can try only **re**placing multiple **consecutive** spaces with `&nbsp;`. so:
```py
re.sub(r' {2,}', '&nbsp', str)
```

Reply:
@rchl correct. The link wont be clickable.

Reply:
We can `split()` on newlines, then find the first non-whitespace char, replace the whitespace prefix with `&nbsp;` ?

And then `"<br>".join(split_lines)` again? :P

Reply:
I suppose the end-all-be-all answer can be deferred for a later PR...

Reply:
I can try something with `re.sub(r' {2,}', replace_npbs, str)`

```
def replace_npbs(match):
   return "&nbsp;" * len(match)
```

Reply:
Replacing only leading spaces would be even easier and likely more efficient with regex. But IMO it's important to also preserve multiple spaces anywhere.
This is maybe crazy but this should work:
```py
re.sub(r' {2,}', lambda match: ''.join(['&nbsp;' for space in match.group(0)]), str)
```

Reply:
Don’t forget tab characters!

Reply:
22d047ebaba5eff3dccade2e32130861d76e5727 and 22d047ebaba5eff3dccade2e32130861d76e5727 :)

Reply:
Tell me if you would like to change anything else. I
I am heading to write some test for plaintext2html.